File: LeetCode/Problem 206.py
# LeetCode: Problem 206
# Jonathan Kosasih

#Reverse a singly linked list.
#
#Example:
#Input: 1->2->3->4->5->NULL
#Output: 5->4->3->2->1->NULL
#------------------------------------------------------------------
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    # initial solution, iterative
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if(not head):
            return None
        scout = head
        current = None
        while(scout):
            temp = scout.next
            scout.next = current
            current = scout
            scout = temp
        return current
    
    # recursive solution
    def reverseListRecursive(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if(head == None or head.next == None):
            return head
        # travel all the way to the end, then start setting stuff
        temp = self.reverseListRecursive(head.next)
        head.next.next = head
        head.next = None
        return temp
    
    def display(self, head):
        while(head):
            print(str(head.val))
            head = head.next

sol = Solution()
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)
node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5
sol.display(node1)
print("------")
sol.display(sol.reverseList(node1))
#sol.display(sol.reverseListRecursive(node1))

File: LeetCode/C++/Problem 206.cpp
// LeetCode: Problem 206
// Jonathan Kosasih
/*
Reverse a singly linked list.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	// iterative
	ListNode* reverseList(ListNode* head) {
		ListNode* scout = head;
		ListNode* current = NULL;
		while (scout) {
			ListNode* temp = scout->next;
			scout->next = current;
			current = scout;
			scout = temp;
		}
		return current;
	}
	// recursive
	ListNode* reverseListRecursive(ListNode* head) {
		if (head == NULL || head->next == NULL) {
			return head;
		}
		ListNode* scout = reverseListRecursive(head->next);
		head->next->next = head;
		head->next = NULL;
		return scout;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}